library/std/src/sys/unix/process/process_unix.rs

```diff
@@ -140,11 +148,32 @@ impl Command {
 
     // Attempts to fork the process. If successful, returns Ok((0, -1))
     // in the child, and Ok((child_pid, -1)) in the parent.
-    #[cfg(not(target_os = "linux"))]
+    #[cfg(not(any(target_os = "linux", target_os = "nto")))]
     unsafe fn do_fork(&mut self) -> Result<(pid_t, pid_t), io::Error> {
         cvt(libc::fork()).map(|res| (res, -1))
     }
 
+    // On QNX Neutrino, fork can fail with EBADF in case "another thread might have opened
+    // or closed a file descriptor while the fork() was occurring".
+    // Documentation says "... or try calling fork() again". This is what we do here.
+    // See also https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/f/fork.html
+    #[cfg(target_os = "nto")]
+    unsafe fn do_fork(&mut self) -> Result<(pid_t, pid_t), io::Error> {
+        use crate::sys::os::errno;
+
+        // arbitrary number of retries, will retry for at least 10 seconds
+        let mut retries = 10_000;
+        loop {
+            let r = libc::fork();
+            if r == -1 as libc::pid_t && retries > 0 && errno() as libc::c_int == libc::EBADF {
+                thread::sleep(Duration::from_millis(1));
+                retries -= 1;
+            } else {
+                return cvt(r).map(|res| (res, -1));
+            }
+        }
```

Comment:
Why are these errors not propagated upwards to be handled, I don't think introducing a sleeping loop is the way to go!

Reply:
Thanks for taking the time for this review!

"Normal" errors from this syscall are propagated upwards. The only exception is `EBADF` where the [documentation](https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/f/fork.html) suggest to try it again:
> A problem occurred when fork() was duplicating a file descriptor. For example, another thread might have opened or closed a file descriptor while the fork() was occurring. You can add synchronization around the operations that involve file descriptors, or **try calling fork() again**.

Only if this fails for a long time (10 seconds), the problem is propagated. Is there a better place to handle this but still be API compatible for users of Rusts stdlib?

Reply:
> The issue here is; it should not ever take _up to_ 10 seconds to find out that a `fork()` (or something to that extent) failed with an `EBADF`. 

This is a good point. We have not been happy with the loop but tried to keep the patch small and the API compatible (users will have to handle this, for example the [remove test server](https://github.com/rust-lang/rust/blob/9b8dbd558c1c4b25c55d987e22baba312ae980ad/src/tools/remote-test-server/src/main.rs#L299)).

> I have not yet read the code surrounding all of this in-depth, and it might be the case that this error cannot - _yet_ - be mapped to some rust equivalent, in this case there should definitely be some way to do this.


> For example we might not be able to fork because there are too many processes already running and returning in this case is absolutely the right thing to do.

At least this particular case would be reported with `EAGAIN` in QNX Neutrino.

If I get it right then the OS error (`EBADF` in this case) is already reported back to the user and can be checked with [raw_os_error](https://doc.rust-lang.org/std/io/struct.Error.html#method.raw_os_error). Wouldn't it be sufficient to remove `nto` specific behavior (and improve handling in e.g. remote test server)?

Reply:
> > The issue here is; it should not ever take _up to_ 10 seconds to find out that a `fork()` (or something to that extent) failed with an `EBADF`.
> 
> This is a good point. We have not been happy with the loop but tried to keep the patch small and the API compatible (users will have to handle this, for example the [remove test server](https://github.com/rust-lang/rust/blob/9b8dbd558c1c4b25c55d987e22baba312ae980ad/src/tools/remote-test-server/src/main.rs#L299)).

I don't get why this would be a problem, the code there already handles errors - doesn't it?

> > I have not yet read the code surrounding all of this in-depth, and it might be the case that this error cannot - _yet_ - be mapped to some rust equivalent, in this case there should definitely be some way to do this.
> 
> > For example we might not be able to fork because there are too many processes already running and returning in this case is absolutely the right thing to do.
> 
> At least this particular case would be reported with `EAGAIN` in QNX Neutrino.

Yes exactly.

> If I get it right then the OS error (`EBADF` in this case) is already reported back to the user and can be checked with [raw_os_error](https://doc.rust-lang.org/std/io/struct.Error.html#method.raw_os_error). Wouldn't it be sufficient to remove `nto` specific behavior (and improve handling in e.g. remote test server)?

I am not sure special handling is even required.

Reply:
> > > The issue here is; it should not ever take _up to_ 10 seconds to find out that a `fork()` (or something to that extent) failed with an `EBADF`.
> > 
> > 
> > This is a good point. We have not been happy with the loop but tried to keep the patch small and the API compatible (users will have to handle this, for example the [remove test server](https://github.com/rust-lang/rust/blob/9b8dbd558c1c4b25c55d987e22baba312ae980ad/src/tools/remote-test-server/src/main.rs#L299)).
> 
> I don't get why this would be a problem, the code there already handles errors - doesn't it?

The remote-test-server currently responses with a panic to any error, i.e. the test run stops and needs to be restarted (in case of remote test server). But this is not the correct handling for this error on QNX Neutrino.
I guess this will be what most applications do.
We would need something like this in remote-test-server (and other apps; quickly hacked for illustration only and won't compile):
```diff
        // Spawn the child and ferry over stdout/stderr to the socket in a framed
        // fashion (poor man's style)
+        #[cfg(not(target_os = "nto"))]
        let mut child =
            t!(cmd.stdin(Stdio::null()).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn());
+        #[cfg(target_os = "nto")]
+       let mut child = {
+           let mut retries = 10_000;
+           let mut r;
+           loop {
+               r = cmd.stdin(Stdio::null()).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn();
+               if r == -1 as libc::pid_t && retries > 0 && errno() as libc::c_int == libc::EBADF {
+                   thread::sleep(Duration::from_millis(1));
+                   retries -= 1;
+               } else {
+                   t!(r)
+               }
+           };
+       };
```

I do see benefits of this solution: The call returns immediately (instead of taking seconds) and the application can decide how many retries make sense (if at all) and if (and how much) sleeping in between makes sense.

On the other hand, it also means that most applications need to implement this same functionality over and over again. This is very QNX Neutrino specific -- in another *nix OS, the return value `libc::EBADF` could have a very different meaning which means also a different handling is required. Having this code in Rusts stdlib would abstract from this specific behavior.

I think this is a bit similar to reading random numbers using `getrandom` which can be interrupted by a signal: In this case, Rusts stdlib just retries (indefinitely) while the return value is `libc::EINTR`):
https://github.com/rust-lang/rust/blob/d1ac43a9b9a8250d858705b0796dfed6186e18db/library/std/src/sys/unix/rand.rs#L98

> 
> I am not sure special handling is even required.

See above.

Reply:
> [...]
> 
> The remote-test-server currently responses with a panic to any error, i.e. the test run stops and needs to be restarted (in case of remote test server). But this is not the correct handling for this error on QNX Neutrino. I guess this will be what most applications do. We would need something like this in remote-test-server (and other apps; quickly hacked for illustration only and won't compile):
> 
> ```diff
>         // Spawn the child and ferry over stdout/stderr to the socket in a framed
>         // fashion (poor man's style)
> +        #[cfg(not(target_os = "nto"))]
>         let mut child =
>             t!(cmd.stdin(Stdio::null()).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn());
> +        #[cfg(target_os = "nto")]
> +       let mut child = {
> +           let mut retries = 10_000;
> +           let mut r;
> +           loop {
> +               r = cmd.stdin(Stdio::null()).stdout(Stdio::piped()).stderr(Stdio::piped()).spawn();
> +               if r == -1 as libc::pid_t && retries > 0 && errno() as libc::c_int == libc::EBADF {
> +                   thread::sleep(Duration::from_millis(1));
> +                   retries -= 1;
> +               } else {
> +                   t!(r)
> +               }
> +           };
> +       };
> ```
> 
> I do see benefits of this solution: The call returns immediately (instead of taking seconds) and the application can decide how many retries make sense (if at all) and if (and how much) sleeping in between makes sense.
> 
> On the other hand, it also means that most applications need to implement this same functionality over and over again. This is very QNX Neutrino specific -- in another *nix OS, the return value `libc::EBADF` could have a very different meaning which means also a different handling is required. Having this code in Rusts stdlib would abstract from this specific behavior.

I don't think this is the right approach either.

The right approach would be to make circumstances such, that it is not possible (or very rarely so) for it to fail in this way.

> I think this is a bit similar to reading random numbers using `getrandom` which can be interrupted by a signal: In this case, Rusts stdlib just retries (indefinitely) while the return value is `libc::EINTR`):
> 
> https://github.com/rust-lang/rust/blob/d1ac43a9b9a8250d858705b0796dfed6186e18db/library/std/src/sys/unix/rand.rs#L98

Well - yes (there's a loop) and no; `getrandom(2)` in case of `EINTR` behaves like `read(2)` where depending on signal handler setup interruptions, need to be handled.

One other suggestion; as we need some special sauce for QNX, why not go all the way and use QNX's [spawn](https://www.qnx.com/developers/docs/7.1/#com.qnx.doc.neutrino.lib_ref/topic/s/spawn.html) directly - it gives all the control one needs in order to prevent these dances around modified FD tables in the first place.

Reply:
@fritschy @gh-tr I've pushed an update:
- Rebased to master (resolving a merge conflict in cargo.lock)
- Removed the controversial loops around fork and spawnp ([ad2b1bf](https://github.com/rust-lang/rust/pull/106673/commits/ad2b1bffe98efe95e41724f45158c5a41e670e5a)). I don't think they are important enough to further delay integration of this PR but could be discussed independently. Problem was only observed when running the test suite with many parallel tests on an aarch64 target and is unlikely to be an issue for anybody else.

When reading gh-tr's comment below another idea occurred to me how we could work around this issue: One could argue that it is a very low-level detail of QNX Neutrino, i.e. below the standard library, at kernel or libc level. The `libc` crate could abstract from this behavior (this would also fix the same issue for other abstractions like the `nix` crate). I don't really like the idea but wanted to bring it up in case it helps.

@fritschy , what do you think? Can we merge this PR as-is and discuss the loop fork/spawnp issue separately, for example in a bug issue?

Reply:
Oh, this looks much better! But I have no authority over merges or PRs, I was just commenting code I though should not go into rust (as it was).

Perhaps @thomcc can take a look.